xcm/pallet-xcm/src/lib.rs

```diff
@@ -448,9 +597,203 @@ pub mod pallet {
+			for v in 0..XCM_VERSION {
+				for (old_key, value) in SupportedVersion::<T>::drain_prefix(v) {
+					if let Ok(new_key) = old_key.into_latest() {
+						SupportedVersion::<T>::insert(XCM_VERSION, new_key, value);
+					}
+				}
+				for (old_key, value) in VersionNotifiers::<T>::drain_prefix(v) {
+					if let Ok(new_key) = old_key.into_latest() {
+						VersionNotifiers::<T>::insert(XCM_VERSION, new_key, value);
+					}
+				}
+			}
+			let xcm_version = T::AdvertizeXcmVersion::get();
+			let response = Response::Version(xcm_version);
+			for (key, mut value) in VersionNotifyTargets::<T>::iter_prefix(XCM_VERSION) {
+				if value.2 == xcm_version {
+					continue
+				}
+				let new_key: MultiLocation = match key.clone().try_into() {
+					Ok(k) => k,
+					Err(_) => continue, // will never happen since it's the latest version already.
+				};
+				value.2 = xcm_version;
+				let message = Xcm(vec![QueryResponse {
+					query_id: value.0,
+					response: response.clone(),
+					max_weight: value.1,
+				}]);
+				match T::XcmRouter::send_xcm(new_key.clone(), message) {
+					Ok(()) => {
+						let versioned_key = LatestVersionedMultiLocation(&new_key);
+						VersionNotifyTargets::<T>::insert(XCM_VERSION, versioned_key, value);
+						Self::deposit_event(Event::VersionChangeNotified(new_key, xcm_version));
+					},
+					Err(e) => {
+						Self::deposit_event(Event::NotifyTargetSendFail(
+							new_key,
+							value.0,
+							e.into(),
+						));
+					},
+				}
+			}
+			for v in 0..XCM_VERSION {
+				for (old_key, mut value) in VersionNotifyTargets::<T>::drain_prefix(v) {
+					let new_key = match MultiLocation::try_from(old_key.clone()) {
+						Ok(k) => k,
+						Err(()) => {
+							Self::deposit_event(Event::NotifyTargetMigrationFail(old_key, value.0));
+							return 0
+						},
+					};
+
+					if value.2 != xcm_version {
+						value.2 = xcm_version;
+						let instruction = QueryResponse {
+							query_id: value.0,
+							response: response.clone(),
+							max_weight: value.1,
+						};
+						match T::XcmRouter::send_xcm(new_key.clone(), Xcm(vec![instruction])) {
+							Ok(()) => (),
+							Err(e) => {
+								let event = Event::NotifyTargetSendFail(new_key, value.0, e.into());
+								Self::deposit_event(event);
+								continue
+							},
+						}
+					}
+					let versioned_key = LatestVersionedMultiLocation(&new_key);
+					VersionNotifyTargets::<T>::insert(XCM_VERSION, versioned_key, value);
+					Self::deposit_event(Event::VersionChangeNotified(new_key, xcm_version));
+				}
+			}
+			0
+		}
+
+		/// Request that `dest` informs us of its version.
+		pub fn request_version_notify(dest: MultiLocation) -> XcmResult {
+			let versioned_dest = VersionedMultiLocation::from(dest.clone());
+			let already = VersionNotifiers::<T>::contains_key(XCM_VERSION, &versioned_dest);
+			ensure!(!already, XcmError::InvalidLocation);
+			let query_id = QueryCount::<T>::mutate(|q| {
+				let r = *q;
+				*q += 1;
```

Comment:
practically speaking, this should be a checked add, and can fail to mutate if overflow

Reply:
Why should that be a *practical* issue?

Reply:
I get that u64 in this context will never overflow, but i just don't see why we wouldn't use checked math here since it comes at practically zero cost. Your call, hopefully its not too much of a pain that I continue to point out such things